llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp

```diff
@@ -4627,10 +4627,15 @@ bool AMDGPUAsmParser::validateDPP(const MCInst &Inst,
     if (Src1Idx >= 0) {
       const MCOperand &Src1 = Inst.getOperand(Src1Idx);
       const MCRegisterInfo *TRI = getContext().getRegisterInfo();
-      if (Src1.isImm() ||
-          (Src1.isReg() && isSGPR(mc2PseudoReg(Src1.getReg()), TRI))) {
-        AMDGPUOperand &Op = ((AMDGPUOperand &)*Operands[Src1Idx]);
-        Error(Op.getStartLoc(), "invalid operand for instruction");
+      if (Src1.isReg() && isSGPR(mc2PseudoReg(Src1.getReg()), TRI)) {
+        auto Reg = mc2PseudoReg(Inst.getOperand(Src1Idx).getReg());
+        SMLoc S = getRegLoc(Reg, Operands);
+        Error(S, "invalid operand for instruction");
+        return false;
+      }
+      if (Src1.isImm()) {
+        Error(getInstLoc(Operands),
+              "src1 immediate operand invalid for instruction");
```

Comment:
Is this part just improving diagnostics, not strictly required for the new feature?

Why do you use different locations for the two errors you produce here?

Reply:
This is necessary because we only expect Src1Idx == the matching idx in Operands if the instruction does not support modifiers. So that was previously true but no longer. For example, we maintain separate lists of the MCInst indices and the parsed operand indices in VOPD. It is actually a downgrade in diagnostic accuracy for immediates, and the same level of precision for sgprs. That is because it is not implemented and would be expensive to determine the position of a certain immediate in the parsed operand list. So we now don't get the '^' pointing to it, but can write in text that src1 is the position with the error.

Reply:
I see - you had to replace the old `Op = ((AMDGPUOperand &)*Operands[Src1Idx])` with something new.